// @iqb/validate-md-profile/schema-validate.factory.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SchemaValidateFactory = void 0;
var ajv_1 = require("ajv");
var profileSchema = "https://raw.githubusercontent.com/iqb-specifications/metadata-profile/refs/heads/main/metadata-profile.schema.json";
var storeSchema = "https://raw.githubusercontent.com/iqb-specifications/metadata-store/refs/heads/main/metadata-store.schema.json";
var SchemaValidateFactory = /** @class */ (function () {
    function SchemaValidateFactory() {
    }
    SchemaValidateFactory.validateProfile = function (sourceFilename) {
        return __awaiter(this, void 0, void 0, function () {
            var mdProfile, fs, compiledSchema, ajv, response, profile, profileData, profile_data_raw, valid, doubleIds_1, allEntryIds_1, entryCount_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        mdProfile = null;
                        fs = require('fs');
                        ajv = new ajv_1.default() // options can be passed, e.g. {allErrors: true}
                        ;
                        return [4 /*yield*/, fetch(profileSchema)];
                    case 1:
                        response = _a.sent();
                        return [4 /*yield*/, response.json()];
                    case 2:
                        profile = _a.sent();
                        try {
                            compiledSchema = ajv.compile(profile);
                        }
                        catch (err) {
                            console.log('\x1b[0;31mERROR\x1b[0m parsing profile schema');
                            console.error(err);
                            process.exitCode = 1;
                            compiledSchema = null;
                        }
                        if (compiledSchema) {
                            if (fs.existsSync(sourceFilename)) {
                                profileData = void 0;
                                try {
                                    profile_data_raw = fs.readFileSync(sourceFilename, 'utf8');
                                    profileData = JSON.parse(profile_data_raw);
                                }
                                catch (err) {
                                    console.log("\u001B[0;31mERROR\u001B[0m reading and parsing profile file '".concat(sourceFilename, "':"));
                                    console.error(err);
                                    profileData = null;
                                    process.exitCode = 1;
                                }
                                if (profileData) {
                                    try {
                                        valid = compiledSchema ? compiledSchema(profileData) : null;
                                        if (!valid) {
                                            console.log("\u001B[0;31mERROR\u001B[0m invalid profile file '".concat(sourceFilename, "':"));
                                            console.error(compiledSchema ? compiledSchema.errors : 'error unknown');
                                            profileData = null;
                                            process.exitCode = 1;
                                        }
                                    }
                                    catch (err) {
                                        console.log("\u001B[0;31mERROR\u001B[0m invalid profile file '".concat(sourceFilename, "':"));
                                        console.error(err);
                                        profileData = null;
                                        process.exitCode = 1;
                                    }
                                    if (profileData) {
                                        try {
                                            mdProfile = {
                                                id: profileData.id,
                                                label: profileData.label,
                                                groups: profileData.groups
                                            };
                                        }
                                        catch (err) {
                                            console.log("\u001B[0;31mERROR\u001B[0m parsing profile file '".concat(sourceFilename, "':"));
                                            console.error(err);
                                            mdProfile = null;
                                            process.exitCode = 1;
                                        }
                                        if (mdProfile) {
                                            doubleIds_1 = [];
                                            allEntryIds_1 = [];
                                            entryCount_1 = 0;
                                            mdProfile.groups.forEach(function (g) {
                                                g.entries.forEach(function (e) {
                                                    if (allEntryIds_1.includes(e.id)) {
                                                        doubleIds_1.push(e.id);
                                                    }
                                                    else {
                                                        allEntryIds_1.push(e.id);
                                                    }
                                                });
                                                entryCount_1 += g.entries.length;
                                            });
                                            console.log("".concat(mdProfile.groups.length, " ").concat(mdProfile.groups.length === 1 ? 'group' : 'groups', " and ").concat(entryCount_1, " ").concat(entryCount_1 === 1 ? 'entry' : 'entries', " found in '").concat(sourceFilename, "'."));
                                            if (doubleIds_1.length > 0) {
                                                console.log("\u001B[0;31mERROR\u001B[0m in profile file '".concat(sourceFilename, "': double ids ").concat(doubleIds_1.join(', ')));
                                                mdProfile = null;
                                                process.exitCode = 1;
                                            }
                                        }
                                    }
                                }
                            }
                            else {
                                console.log("\u001B[0;31mERROR\u001B[0m profile file '".concat(sourceFilename, "' not found"));
                                process.exitCode = 1;
                            }
                        }
                        return [2 /*return*/, mdProfile];
                }
            });
        });
    };
    SchemaValidateFactory.validateConfig = function (sourceFilename) {
        return __awaiter(this, void 0, void 0, function () {
            var mdStore, fs, compiledSchema, ajv, response, store, profileStoreData, config_data_raw, valid;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        mdStore = null;
                        fs = require('fs');
                        ajv = new ajv_1.default() // options can be passed, e.g. {allErrors: true}
                        ;
                        return [4 /*yield*/, fetch(storeSchema)];
                    case 1:
                        response = _a.sent();
                        return [4 /*yield*/, response.json()];
                    case 2:
                        store = _a.sent();
                        try {
                            compiledSchema = ajv.compile(store);
                        }
                        catch (err) {
                            console.log('\x1b[0;31mERROR\x1b[0m parsing profile config schema');
                            console.error(err);
                            process.exitCode = 1;
                            compiledSchema = null;
                        }
                        if (compiledSchema) {
                            if (fs.existsSync(sourceFilename)) {
                                profileStoreData = void 0;
                                try {
                                    config_data_raw = fs.readFileSync(sourceFilename, 'utf8');
                                    profileStoreData = JSON.parse(config_data_raw);
                                }
                                catch (err) {
                                    console.log("\u001B[0;31mERROR\u001B[0m reading and parsing config file '".concat(sourceFilename, "':"));
                                    console.error(err);
                                    profileStoreData = null;
                                    process.exitCode = 1;
                                }
                                if (profileStoreData) {
                                    try {
                                        valid = compiledSchema ? compiledSchema(profileStoreData) : null;
                                        if (!valid) {
                                            console.log("\u001B[0;31mERROR\u001B[0m invalid config file '".concat(sourceFilename, "':"));
                                            console.error(compiledSchema ? compiledSchema.errors : 'error unknown');
                                            profileStoreData = null;
                                            process.exitCode = 1;
                                        }
                                    }
                                    catch (err) {
                                        console.log("\u001B[0;31mERROR\u001B[0m invalid config file '".concat(sourceFilename, "':"));
                                        console.error(err);
                                        profileStoreData = null;
                                        process.exitCode = 1;
                                    }
                                    if (profileStoreData) {
                                        try {
                                            mdStore = {
                                                id: profileStoreData.id,
                                                title: profileStoreData.title,
                                                creator: profileStoreData.creator,
                                                maintainer: profileStoreData.maintainer,
                                                profiles: profileStoreData.profiles
                                            };
                                        }
                                        catch (err) {
                                            console.log("\u001B[0;31mERROR\u001B[0m instanciating profile store '".concat(sourceFilename, "':"));
                                            console.error(err);
                                            mdStore = null;
                                            process.exitCode = 1;
                                        }
                                    }
                                }
                            }
                            else {
                                console.log("\u001B[0;31mERROR\u001B[0m config file '".concat(sourceFilename, "' not found"));
                                process.exitCode = 1;
                            }
                        }
                        return [2 /*return*/, mdStore];
                }
            });
        });
    };
    return SchemaValidateFactory;
}());
exports.SchemaValidateFactory = SchemaValidateFactory;
